Upgrade from boto to boto3, and use env variables

GitRon opened the pull request:
Hi guys,

I started the migration from `boto` to `boto3`.

I tried it out locally and now I can send from an email region which has this new auth type v4.

**Here's a list of stuff I already did:**
* Upgrade from boto to boto3
* Some flake8 linting
* Fixed both management command
* Sending mails from eu-central-region works now
* Removed proxy and updated docs

Fixes #181 #78

pcraciunoiu commented:
@GitRon thanks for starting on this! The build fails, would be good to fix.

Could you post this in the relevant project issues? That way others can find it and try to contribute as well.

GitRon commented:
> 
> 
> @GitRon thanks for starting on this! The build fails, would be good to fix.
> 
> Could you post this in the relevant project issues? That way others can find it and try to contribute as well.

I added a list in the PR to all known issues we have to deal with. I'll have a look at the tests, at least this is my responsibility ;)

And just got the info that the issue got a new level of urgency: **To improve the security for our customers, beginning October 1, 2020, Amazon Signature Version 3 will be turned off (deprecated) in Amazon SES in favor of Signature Version 4.**

trecouvr commented on setup.py:
```diff
@@ -131,9 +129,10 @@ def find_package_data(where=".", package="", exclude=standard_exclude,
+    version='0.9.0',  # When changing this, remember to change it in __init__.py
     packages=find_packages(exclude=['example', 'tests']),
     package_data=package_data,
+    python_requires='>=3.5.8',
```
```suggestion
```

GitRon replied:
thanks so much for your input! I changed the python version to `3.4.10` and removed python 2.7 in the travix matrix. After all, python 2.7 is dead. Long live python 3.

pcraciunoiu replied:
I'm fine dropping support for python 2, though if it's not a big deal we could keep it (I suspect some users will be affected). However it's not worth a lot of time... And hopefully not too many users complain.

It would be great if you could add the supported Django versions to the README. As far as that, I'd be fine supporting 1.10 or even 1.11 and above. Django 3 should be easy if 2.2 works...

GitRon replied:
Ok, when the tests are fixed, I'll give Python 2.7 another shot and I added django 1.10 to the tox and travis.

trecouvr commented on setup.py:
```diff
@@ -142,7 +141,7 @@ def find_package_data(where=".", package="", exclude=standard_exclude,
     platforms=['any'],
     classifiers=CLASSIFIERS,
-    install_requires=["boto>=2.31.0", "pytz>=2016.10", "future>=0.16.0"],
+    install_requires=["boto3>=1.0.0", "pytz>=2016.10", "future>=0.16.0", "django>1.11"],
```
```suggestion
    install_requires=["boto3>=1.0.0", "pytz>=2016.10", "future>=0.16.0"],
```

GitRon replied:
Why did you remove the django dependency here?

trecouvr replied:
I removed it because because Travis matrix has Django 1.11 (https://github.com/django-ses/django-ses/blob/master/.travis.yml#L12) and because it was not on the initial source (https://github.com/django-ses/django-ses/blob/master/setup.py#L145), but maybe it's a good thing to add it?

GitRon replied:
I mainly added it, because the IDE gives you loades of warnings if it's missing 😅

GitRon commented:
> 
> 
> If we do keep support for python 2.7, I wonder if it would be easy to add a deprecation warning for it, saying this will be the last release to support it. See this article for a bit on warnings if you're new to them: https://lerner.co.il/2020/04/27/working-with-warnings-in-python/

Seems like a good idea. I'll look into it when we know for sure that we'll stick to python 2.7

GitRon commented:
We found [this post](https://stackoverflow.com/a/33501223/1331671) stating we don't need a proxy anymore (which has been removed in boto3 anyway) and you can do it via the ENV vars. Anybody disagree if we remove it?

pcraciunoiu commented:
Yeah that seems fine to remove if it can work via ENV vars.

chrisconlan commented:
Following this because of the Signature V4 deadline. Thanks for all your great work!

GitRon commented:
> 
> 
> Yeah that seems fine to remove if it can work via ENV vars.
> 
> Though if it is a separate change, you can also do it in the future/separate PR. Whichever is easier?

I removed the proxy logic and added a part to the documentation.

GitRon commented:
Ok, update on my progress: dkim is back in, the package seems not to be related to the boto3 upgrade. Though I cannot test it, it still just keep working.

Still struggling with some tests, working on mocking the `send_raw_mail` function properly... That's my current task.

trecouvr commented:
Hi,

I fixed the test based on GitRon work. Everything passes except the Python 2.7 tests because of the requirements : https://travis-ci.org/github/django-ses/django-ses/builds/696453533.

Also I used a different approach than yours GitRon to mock the different classes, this result in less code diff between the PR and the current master.

Hope this solves most of issues.

Best,

GitRon commented:
> 
> 
> Hi,
> 
> I fixed the tests based on GitRon work. Everything passes except the Python 2.7 tests because of the requirements : https://travis-ci.org/github/django-ses/django-ses/builds/696453533.
> 
> Also I used a different approach than yours GitRon to mock the different classes, this result in less code diff between the PR and the current master.
> 
> Hope this solves most of issues. I made a PR to your repo GitRon, so you can integrate my changes with just one click ;).
> 
> Best,

@trecouvr You are the best! ❤️  I have some questions about your fix - if you might clarfiy it, then I'll merge it gratefully into my branch!

GitRon commented:
@pcraciunoiu Thanks to @trecouvr the tests are running, I just added some basic `flake8` linting and then we are ready for a review I'd say 😃

GitRon commented:
There was a mixup with the merge, I pushed a fix.

trecouvr commented on setup.py:
```diff
@@ -70,22 +71,22 @@ def find_package_data(where=".", package="", exclude=standard_exclude,
+                            fn.lower() == pattern.lower():
                         bad_name = True
                         if show_ignored:
                             print >> sys.stderr, (
```
It's python 2 syntax I think.

GitRon replied:
I replaced it with a nicer solution from stack with 500+ votes anyway 😃

trecouvr commented:
Hi, I'll make some time to test it through our applications using it. But we only use the send part of this library, never used a proxy, dkim or the stats.

trecouvr commented:
Maybe the PR https://github.com/django-ses/django-ses/pull/73 could be merged, then we can ask @brutasse to re-test the dkim part with this branch?

GitRon commented:
> 
> 
> @GitRon sounds good! Thanks for doing this!!

I successfully sent emails via eu-west and eu-central - so I guess it works now 🚀

GitRon commented:
I like the idea of getting dkim fixed as well but the branch is quite old and we need the base functionality going. so if we don't get a reply soon, I suggest we throw out version 1.0.0 and can always add new bugfix releases if some of the minor functionality is broken. what do you think about this approach?

trecouvr commented on django_ses/management/commands/get_ses_statistics.py:
```diff
@@ -29,13 +26,10 @@ class Command(BaseCommand):
+        connection = boto3.client(
+            'ses',
             aws_access_key_id=settings.ACCESS_KEY,
             aws_secret_access_key=settings.SECRET_KEY,
```
I had to add these lines to make it work on my environment (copied from the other command `ses_email_address.py`), but I dont think this is related to the current on going work. My problem is, by default I use a region which does not have SES, thus I have to manually chose the ses region to work on.

```python
            region_name=settings.AWS_SES_REGION_NAME,
            endpoint_url=settings.AWS_SES_REGION_ENDPOINT_URL,
```

GitRon replied:
@trecouvr Shall I add these two lines to the code?

trecouvr replied:
I think you should, these are the same lines as in the other parts of the code. It allows to specify the ses region.

I would say the best move would be to refactor all the boto.client calls present in the lib and in the commands into one function, but this could be done later.

GitRon replied:
Pushed it. Thanks for the review. And yes, we definetely should refactor it at some point.

trecouvr commented:
Ok I managed to send emails and get the 2 managements commands working (needed a change for the statistics, but I dont think this is related to the ongoing work, see my comment above).

As for the DKIM part, not sure how many people use it, I use what you call in your doc "Easy DKIM" which allows SES to sign directly the emails.

pcraciunoiu commented:
@GitRon @trecouvr so does this seem ready from your end? I might wait until Monday or do a release on the weekend, but wanted to confirm it's good to go. I'll try and re-review soon too, but I imagine it's in good shape now :)

GitRon commented:
@pcraciunoiu The stuff I need now works. About all the things which are included in this package... maybe there should be a reevaluation, if we still need to carry it all around. Like dkim, I think there is a AWS way of doing it.
Really looking forward to the release! 🎉

pcraciunoiu commented on tests/test_commands.py:
```diff
@@ -1,60 +1,53 @@
     {
         'Complaints': '1',
-        'Timestamp': '2012-01-01T02:00:00Z',
+        'Timestamp': datetime.datetime(year=2012, month=1, day=1, hour=2),
```
@GitRon @trecouvr has anyone tested this manually? Just want to be sure it at least works to run the stats and the returned data has this shape still... I'd assume so, but it would likely come up as a regression if not.

GitRon replied:
About 2 weeks ago I tested the statistics command and realised that it now returns a date object rather than a string. So it should work like this 😅

pcraciunoiu commented on django_ses/__init__.py:
```diff
@@ -193,28 +185,29 @@ def send_messages(self, email_messages):
-                                          dkim_headers=self.dkim_headers)
+                    Source=source or message.from_email,
+                    Destinations=message.recipients(),
+                    # todo attachments?
```
@GitRon were you planning on doing this, is it worth highlighting in a changelog?

GitRon replied:
Good point! I actually wanted to test attachements to make sure they would work but totally forgot about it! I'll check it asap.

pcraciunoiu commented:
Made a couple small changes (the version parsing wasn't quite right -- needed to be joined into a string).

Release created: https://github.com/django-ses/django-ses/releases/tag/v1.0.0

And pushed to pypi: https://pypi.org/project/django-ses/1.0.0/

Please test it out. Can do a patch release if stats is broken.

pcraciunoiu commented:
> @pcraciunoiu The stuff I need now works. About all the things which are included in this package... maybe there should be a reevaluation, if we still need to carry it all around. Like dkim, I think there is a AWS way of doing it.

Yeah I would be glad to remove stuff that is supported in different ways.

GitRon commented:
> 
> 
> > @pcraciunoiu The stuff I need now works. About all the things which are included in this package... maybe there should be a reevaluation, if we still need to carry it all around. Like dkim, I think there is a AWS way of doing it.
> 
> Yeah I would be glad to remove stuff that is supported in different ways.

Maybe we could open an issue for it? So it won't be forgotten?

pcraciunoiu commented:
> > > @pcraciunoiu The stuff I need now works. About all the things which are included in this package... maybe there should be a reevaluation, if we still need to carry it all around. Like dkim, I think there is a AWS way of doing it.
> > 
> > 
> > Yeah I would be glad to remove stuff that is supported in different ways.
> 
> Maybe we could open an issue for it? So it won't be forgotten?

Yeah want to go ahead and open one, listing what you have in mind?

GitRon commented:
> 
> 
> > > > @pcraciunoiu The stuff I need now works. About all the things which are included in this package... maybe there should be a reevaluation, if we still need to carry it all around. Like dkim, I think there is a AWS way of doing it.
> > > 
> > > 
> > > Yeah I would be glad to remove stuff that is supported in different ways.
> > 
> > 
> > Maybe we could open an issue for it? So it won't be forgotten?
> 
> Yeah want to go ahead and open one, listing what you have in mind?

Nothing in particular, just maybe look at all the functionality and management commands I personally never use and check if they a) work and b) are still required.

I just tested sending emails with attachments - it works! So my todo was unnecessary!
